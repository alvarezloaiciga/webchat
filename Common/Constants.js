export const SupportedWebchatUrls = [
  'goquiq.com/app/webchat',
  'quiq-api.com/app/webchat',
  'quiq.sh/app/webchat',
  'centricient.com/app/webchat',
  'centricient.corp/app/webchat',
  'quiq.dev:7000/app/webchat',
  'quiq.dev:3000/app/webchat',
  'centricient.dev:7000/app/webchat',
  'quiq.dev:41014/app/webchat',
  'centricient.dev:41014/app/webchat',
];

export const unknownErrorMessage = 'An unknown error occurred.';

// Default launch button classes and id's
export const noAgentsAvailableClass = 'noAgentsAvailable';
export const mobileClass = 'mobile';
export const unsupportedBrowserClass = 'unsupportedBrowser';
export const storageDisabledClass = 'storageDisabled';
export const hasMobileNumberClass = 'hasMobileNumber';

export const chatClosedClass = 'ChatClosed';
export const chatOpenClass = 'ChatOpen';
export const quiqContainerId = 'quiqWebChat';

export const webchatPath = 'app/webchat/webchat.html';
export const bridgePath = 'app/webchat/bridge.html';

export const StandaloneWindowName = 'quiq-standalone-webchat';

export const eventTypes: {[string]: string} = {
  messageArrived: 'QUIQ_MESSAGE_ARRIVED',
  chatVisibilityDidChange: 'QUIQ_CHAT_VISIBILITY_DID_CHANGE',
  _standaloneOpen: 'QUIQ_STANDALONE_OPEN',
  _launchButtonVisibilityShouldChange: 'QUIQ_LAUNCH_BUTTON_VISIBILITY_SHOULD_CHANGE',
  _configurationDidChange: 'QUIQ_CONFIGURATION_DID_CHANGE',
  _chatIsReady: 'QUIQ_CHAT_IS_READY',
};

// Any eventType that does not begin with "_" is exposed on the Quiq object handed to the client.
export const publicEventTypes = Object.keys(eventTypes).reduce(
  (acc, k) => Object.assign({}, acc, k.startsWith('_') ? {} : {[k]: eventTypes[k]}),
  {},
);

export const MessageTypes = {
  TEXT: 'Text',
  ATTACHMENT: 'Attachment',
};

export const MessageStatus = {
  PENDING: 'pending',
  FAILED: 'failed',
};

export const Sentry = {
  endpoint:
    'https://sentry.io/api/170102/store/?sentry_key=5622397f17c44165a039d3b91d3e0193&sentry_version=7&sentry_client=Direct',
};

export const EventTypes = {
  JOIN: 'Join',
  LEAVE: 'Leave',
  REGISTER: 'Register',
  AGENT_TYPING: 'AgentTyping',
  AGENT_ENDED_CONVERSATION: 'AgentEndedConversation',
  END: 'End',
  SPAM: 'Spam',
  SEND_TRANSCRIPT: 'SendTranscript',
  ASSIGNED: 'Assigned',
};

export const AuthorTypes = {
  CUSTOMER: 'Customer',
  USER: 'User',
  SYSTEM: 'System',
};

export const AttachmentErrorTypes = {
  TOO_LARGE: 'attachmentTooLarge',
  UNSUPPORTED_TYPE: 'attachmentUnsupportedType',
  UPLOAD_ERROR: 'attachmentUploadError',
  EMPTY: 'emptyUpload',
};

// Define subsets of events
export const EndEventTypes = [EventTypes.END, EventTypes.SPAM];

export const DisplayElementTypes = [
  MessageTypes.TEXT,
  MessageTypes.ATTACHMENT,
  EventTypes.SEND_TRANSCRIPT,
  EventTypes.END,
  EventTypes.SPAM,
  EventTypes.ASSIGNED,
  AttachmentErrorTypes.TOO_LARGE,
  AttachmentErrorTypes.UNSUPPORTED_TYPE,
  AttachmentErrorTypes.UPLOAD_ERROR,
  AttachmentErrorTypes.EMPTY,
];

export const reduxActionTypes = {
  uploadProgress: 'UPLOAD_PROGRESS',
  addPendingMessage: 'ADD_PENDING_MESSAGE',
  updatePendingMessageId: 'UPDATE_PENDING_MESSAGE_ID',
  updatePlatformEvents: 'UPDATE_PLATFORM_EVENTS',
  configurationLoaded: 'CHAT_CONFIGURATION_LOADED',
  removeMessage: 'REMOVE_MESSAGE',
  isAgentAssigned: 'AGENT_ASSIGNED',
  addAttachmentError: 'ADD_ATTACHMENT_ERROR',
  markSendFailure: 'MARK_SEND_FAILURE',
};

export const postmasterActionTypes = {
  loadChat: 'QUIQ_LOAD_CHAT',
  unloadChat: 'QUIQ_UNLOAD_CHAT',
  getChatStatus: 'QUIQ_GET_CHAT_STATUS',
  setChatRegistrationField: 'QUIQ_SET_CHAT_REGISTRATION_FIELD',
  setChatVisibility: 'QUIQ_SET_CHAT_VISIBILITY',
  getChatVisibility: 'QUIQ_GET_CHAT_VISIBILITY',
  getAgentAvailability: 'QUIQ_GET_AGENT_AVAILABILITY',
  getHandle: 'QUIQ_GET_HANDLE',
  login: 'QUIQ_LOGIN',
  getCanFlashNotifications: 'QUIQ_GET_CAN_FLASH_NOTIFICATIONS',
  getMobileChatEnabled: 'QUIQ_GET_MOBILE_CHAT',
  getLocalStorage: 'QUIQ_GET_LOCAL_STORAGE',
};

export const ExtensionSdkEventTypes = {
  ESTIMATED_WAIT_TIME_CHANGED: 'estimatedWaitTimeChanged',
};

export const ChatInitializedState = {
  UNINITIALIZED: 'uninitialized',
  LOADING: 'loading',
  INITIALIZED: 'initialized',
  ERROR: 'error',
  DISCONNECTED: 'disconnected',
  BURNED: 'burned',
  SLEEPING: 'sleeping',
};

export const intlMessageTypes = {
  pageTitle: 'pageTitle',
  titleText: 'titleText',
  headerText: 'headerText',
  sendButtonLabel: 'sendButtonLabel',
  messageFieldPlaceholder: 'messageFieldPlaceholder',
  invalidEmailErrorMessage: 'invalidEmailErrorMessage',
  welcomeFormValidationErrorMessage: 'welcomeFormValidationErrorMessage',
  welcomeFormSubmitButtonLabel: 'welcomeFormSubmitButtonLabel',
  welcomeFormSubmittingButtonLabel: 'welcomeFormSubmittingButtonLabel',
  connectingMessage: 'connectingMessage',
  reconnectingMessage: 'reconnectingMessage',
  agentTypingMessage: 'agentTypingMessage',
  agentAssignedMessage: 'agentAssignedMessage',
  agentsNotAvailableMessage: 'agentsNotAvailableMessage',
  agentEndedConversationMessage: 'agentEndedConversationMessage',
  errorMessage: 'errorMessage',
  requiredFieldAriaLabel: 'requiredFieldAriaLabel',
  dockWindowTooltip: 'dockWindowTooltip',
  openInNewWindowTooltip: 'openInNewWindowTooltip',
  optionsMenuTooltip: 'optionsMenuTooltip',
  emailTranscriptMenuMessage: 'emailTranscriptMenuMessage',
  emailTranscriptMenuTooltip: 'emailTranscriptMenuTooltip',
  emailTranscriptInputPlaceholder: 'emailTranscriptInputPlaceholder',
  emailTranscriptInputCancelTooltip: 'emailTranscriptInputCancelTooltip',
  emailTranscriptInputSubmitTooltip: 'emailTranscriptInputSubmitTooltip',
  emailTranscriptInlineButton: 'emailTranscriptInlineButton',
  messageArrivedNotification: 'messageArrivedNotification',
  closeWindowTooltip: 'closeWindowTooltip',
  emojiPickerTooltip: 'emojiPickerTooltip',
  attachmentBtnTooltip: 'attachmentBtnTooltip',
  unsupportedFileType: 'unsupportedFileType',
  attachmentTooLarge: 'attachmentTooLarge',
  attachmentUploadError: 'attachmentUploadError',
  muteSounds: 'muteSounds',
  unmuteSounds: 'unmuteSounds',
  muteSoundsTooltip: 'muteSoundsTooltip',
  unmuteSoundsTooltip: 'unmuteSoundsTooltip',
  transcriptEmailedEventMessage: 'transcriptEmailedEventMessage',
  cannotStartNewConversationMessage: 'cannotStartNewConversationMessage',
  unsupportedOrientation: 'unsupportedOrientation',
  unableToSend: 'unableToSend',
  infectedFile: 'infectedFile',
  emptyUpload: 'emptyUpload',
};

export const localStorageKeys = ['quiq-data'];

export const MenuItemKeys = {
  EMAIL_TRANSCRIPT: 'emailTranscript',
  MUTE_SOUNDS: 'muteSounds',
};

export const UserEmailKey = 'quiq-client-data';

export const maxAttachmentSize = 25 * 1000000;

export const displayModes = {
  DOCKED: 'docked',
  UNDOCKED: 'undocked',
  EITHER: 'either',
};
