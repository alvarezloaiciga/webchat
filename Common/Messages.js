import {defineMessages} from 'react-intl';

export default defineMessages({
  messageArrivedNotification: {
    id: 'messageArrivedNotification',
    description: 'Message Arrived',
    defaultMessage: 'New Message from Quiq Webchat',
  },
  send: {
    id: 'send',
    description: 'Generic send message',
    defaultMessage: 'Send',
  },
  sendUsAMessage: {
    id: 'sendUsAMessage',
    description: 'Generic Send us a message',
    defaultMessage: 'Send us a message...',
  },
  reconnecting: {
    id: 'reconnecting',
    description: 'Reconnecting to websocket',
    defaultMessage: 'Reconnecting...',
  },
  connecting: {
    id: 'connecting',
    description: 'Connecting to websocket',
    defaultMessage: 'Connecting...',
  },
  errorState: {
    id: 'errorState',
    description: 'Generic Error Message',
    defaultMessage: "We're having trouble connecting. Try refreshing the page.",
  },
  clientInactive: {
    id: 'clientInactive',
    description: 'Generic Client Inactive Message',
    defaultMessage:
      'You have been disconnected due to inactivity. Please refresh the page to reconnect.',
  },
  agentIsTyping: {
    id: 'agentIsTyping',
    description: 'Message to display when the agent is typing',
    defaultMessage: 'Agent is typing',
  },
  agentEndedConversation: {
    id: 'agentEndedConversation',
    description: 'Message to display when the agent ends conversation',
    defaultMessage: 'Agent has ended the conversation.',
  },
  agentsNotAvailable: {
    id: 'agentsNotAvailable',
    description: 'Message to display when no agents are avialable.',
    defaultMessage: 'No agents are currently available.',
  },
  required: {
    id: 'required',
    description: 'Required',
    defaultMessage: 'Required',
  },
  referrer: {
    id: 'referrer',
    description: 'Message to append to welcome form telling agent what web page the customer is on',
    defaultMessage: 'Referrer',
  },
  cannotFindScript: {
    id: 'cannotFindScript',
    description: "Error to display when we can't find the script that loaded webchat",
    defaultMessage: 'Cannot find script that loaded Webchat. Please contact your administrator.',
  },
  standaloneFatalError: {
    id: 'standaloneFatalError',
    description:
      'Error to display when we load the quiqOptions object from query params in standalone mode',
    defaultMessage: 'Error loading standalone mode. Are the query params properly formatted?',
  },
  errorParsingStandaloneObject: {
    id: 'errorParsingStandaloneObject',
    description:
      "Error to display when we found the quiqOptions query param but weren't able to parse it into a valid quiqOptions object",
    defaultMessage: 'Error while trying to parse the standalone quiqOptions query object',
  },
  cannotDetermineHost: {
    id: 'cannotDetermineHost',
    description:
      "Error to display when we can't determine the hostid from the script src attribute",
    defaultMessage: 'Cannot determine host from script url. Please contact your administrator',
  },
  quiqFatalError: {
    id: 'quiqFatalError',
    description: 'Message to show a quiq fatal error',
    defaultMessage: 'quiqOptions FATAL ERROR',
  },
  hereToHelp: {
    id: 'hereToHelp',
    description: 'Welcome message to display at top of webchat',
    defaultMessage: "We're here to help if you have any questions!",
  },
  openInNewWindow: {
    id: 'openInNewWindow',
    description: 'Message to display on open in new window icon',
    defaultMessage: 'Open chat in new window',
  },
  emojiPickerTooltip: {
    id: 'emojiPickerTooltip',
    description: 'Tooltip for emoji picker toggle button',
    defaultMessage: 'Emoji picker',
  },
  optionsMenuTooltip: {
    id: 'optionsMenuTooltip',
    description: 'Tooltip for the options button to the left of the MessageForm',
    defaultMessage: 'Options',
  },
  emailTranscriptMenuMessage: {
    id: 'emailTranscriptMenuMessage',
    description:
      'Message to display in the options menu for the user to get an email containing the transcript of the chat',
    defaultMessage: 'Email Transcript',
  },
  emailTranscriptMenuTooltip: {
    id: 'emailTranscriptMenuTooltip',
    description: 'Tooltip for Email transcript menu button',
    defaultMessage: 'Email a full transcript of the current chat',
  },
  emailTranscriptInputPlaceholder: {
    id: 'emailTranscriptInputPlaceholder',
    description:
      'Message to display to user when they need to enter an Email to receive their transcript',
    defaultMessage: 'Enter your Email...',
  },
  emailTranscriptInputCancelTooltip: {
    id: 'emailTranscriptInputCancelTooltip',
    description: 'Tooltip to display above the Cancel button on the email transcript input',
    defaultMessage: 'Cancel Email Transcript',
  },
  emailTranscriptInputSubmitTooltip: {
    id: 'emailTranscriptInputSubmitTooltip',
    description: 'Tooltip to display above the Submit button on the email transcript input',
    defaultMessage: 'Email Transcript',
  },
  emailTranscriptInlineButton: {
    id: 'emailTranscriptInlineButton',
    description:
      'Text to display on the Email Transcript button that shows inline when agent ends convo',
    defaultMessage: 'Email Transcript',
  },
  transcriptEmailedEventMessage: {
    id: 'transcriptEmailedEventMessage',
    description:
      'Text to display on the inline Event displayed to the user when a transcript was successfully e-mailed to them',
    defaultMessage: 'Transcript Emailed',
  },
  attachmentBtnTooltip: {
    id: 'attachmentBtnTooltip',
    description: 'Tooltip for attachment button',
    defaultMessage: 'Send file',
  },
  welcomeFormUniqueIdentifier: {
    id: 'wecomeFormUniqueIdentifier',
    description:
      'Unique identifier to key off of as identifying a message as the welcome form customer submission.',
    defaultMessage: 'Quiq Welcome Form Customer Submission',
  },
  dockWindow: {
    id: 'dockWindow',
    description:
      'Message to display when user hovers over button in standalone mode to dock the standalone window back into the parent window',
    defaultMessage: 'Dock chat',
  },
  closeWindow: {
    id: 'closeWindow',
    description: 'Close Window button',
    defaultMessage: 'Close window',
  },
  minimizeWindow: {
    id: 'minimizeWindow',
    description: 'Minimize Window button',
    defaultMessage: 'Minimize window',
  },
  cookiesMustBeEnabledError: {
    id: 'cookiesMustBeEnabledError',
    description: 'Error message to display when end user does not have cookies enabled',
    defaultMessage: 'Cookies must be enabled to use Webchat',
  },
  invalidWelcomeFormArray: {
    id: 'invalidWelcomeFormArray',
    description:
      'Error message to display when WELCOME_FORM is not valid because fields must be of type array.',
    defaultMessage: 'Invalid WELCOME_FORM definition: "WELCOME_FORM.fields" must be an array',
  },
  invalidWelcomeFormUndefined: {
    id: 'invalidWelcomeFormUndefined',
    description:
      'Error message to display when WELCOME_FORM is not valid because a field is missing required property.',
    defaultMessage:
      'Invalid WELCOME_FORM definition: The form field with id of "{id}" and label of "{label}" must have id, label and type defined.',
  },
  invalidWelcomeFormDefinitionKeyLength: {
    id: 'invalidWelcomeFormDefinitionKeyLength',
    description:
      'Error message to display when WELCOME_FORM is not valid because a field has an id that is too long.',
    defaultMessage:
      'Invalid WELCOME_FORM definition: The form field with id of "{id}" must have an id of length less or equal to than 80 characters.',
  },
  invalidWelcomeFormDefinitionKeyUniqueness: {
    id: 'invalidWelcomeFormDefinitionKeyUniqueness',
    description:
      "Error message to display when WELCOME_FORM is not valid because not all fields have unique id's.",
    defaultMessage: 'Invalid WELCOME_FORM definition: All fields must have a unique id',
  },
  invalidWelcomeFormDefinitionKeyReserved: {
    id: 'invalidWelcomeFormDefinitionKeyReserved',
    description:
      'Error message to display when WELCOME_FORM is not valid because a field has an id that is reserved',
    defaultMessage:
      'Invalid WELCOME_FORM definition: Fields cannot have an id of "{id}" as this id is used internally.',
  },
  invalidWelcomeFormFieldCount: {
    id: 'invalidWelcomeFormFieldCount',
    description:
      'Error message to display when WELCOME_FORM is not valid because there are too many fields',
    defaultMessage: 'Invalid WELCOME_FORM definition: There can be at most 20 fields',
  },
  invalidWelcomeFormFieldType: {
    id: 'invalidWelcomeFormFieldType',
    description:
      'Error message to display when WELCOME_FORM is not valid because an unsupported field type was given',
    defaultMessage: 'Invalid WELCOME_FORM definition: "{type}" is not a supported field type.',
  },
  invalidWelcomeFormFieldRowsType: {
    id: 'invalidWelcomeFormFieldRowsType',
    description:
      'Error message to display when WELCOME_FORM is not valid because a non-numeric value was specified for a field\'s "rows" parameter',
    defaultMessage:
      'Invalid WELCOME_FORM definition: All fields\' "rows" parameters must be of type number',
  },
  submitWelcomeForm: {
    id: 'submitWelcomeForm',
    description: 'Text of submit button on welcome form',
    defaultMessage: 'Submit',
  },
  submittingWelcomeForm: {
    id: 'submittingWelcomeForm',
    description: 'Text of submit button on welcome form while form is being submitted',
    defaultMessage: 'Submitting',
  },
  welcomeFormValidationError: {
    id: 'welcomeFormValidationError',
    description: 'Error displayed when user has not filled in a required field',
    defaultMessage: 'Please complete all fields marked with an *',
  },
  unableToBindCustomLauncherError: {
    id: 'unableToBindCustomLauncherError',
    description: 'Error to display when CUSTOM_LAUNCH_BUTTONS is misconfigured',
    defaultMessage: 'Unable to bind custom webchat launch buttons',
  },
  unableToFindCustomLauncherError: {
    id: 'unableToFindCustomLauncherError',
    description:
      'Error to display once we successfully bind the webchat buttons, but they are now missing',
    defaultMessage: 'Unable to find already-bound custom webchat launch buttons',
  },
  cssHttpsError: {
    id: 'cssHttpsError',
    description: 'Error displayed when css file is not an https url',
    defaultMessage: 'The CUSTOM_CSS_URL must be an HTTPS url.',
  },
  mfInitNeeded: {
    id: 'mfInitNeeded',
    description:
      "Error displayed when init() has not been run in malfunction junction and we try and run something we shouldn't",
    defaultMessage:
      'MalfunctionJunction.init() must be run before posting a message, setting up listeners, or setting up Redux observers.',
  },
  cannotFindHostingWindow: {
    id: 'cannotFindHostingWindow',
    description: "Error for when webchat can't find the host window (where SDK is running)",
    defaultMessage: 'Unable to find host window',
  },
  invalidAttachmentMessage: {
    id: 'invalidAttachmentMessage',
    description: 'Alert message for invalid attachment upload',
    defaultMessage:
      "We couldn't send your attachment. Only image files are supported, and must be smaller than 50 MB",
  },
  muteSounds: {
    id: 'muteSounds',
    description: 'label for muting sounds',
    defaultMessage: 'Mute Sounds',
  },
  unmuteSounds: {
    id: 'unmuteSounds',
    description: 'label for unmuting sounds',
    defaultMessage: 'Unmute Sounds',
  },
  muteSoundsTooltip: {
    id: 'muteSoundsTooltip',
    description: 'tooltip for muting sounds',
    defaultMessage: 'Mute new message sounds',
  },
  unmuteSoundsTooltip: {
    id: 'unmuteSoundsTooltip',
    description: 'tooltip for unmuting sounds',
    defaultMessage: 'Unmute new message sounds',
  },
});
