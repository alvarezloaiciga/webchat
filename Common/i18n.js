// @flow
import type {IntlObject, IntlMessage} from './types';

let intl;

/**
 * Registers the intl object exposed in injectIntl so that it can be used elsewhere
 */
export const registerIntlObject = (intlObject: IntlObject): void => {
  intl = intlObject;
};

/**
 * Passes values to the formatMessage function of the intl object
 * @param {object} message - The react-intl message to translate
 * @param {object=} values - An object of values to pass into formatMessage
 */
export const formatMessage = (message: IntlMessage, values: Object = {}): string => {
  if (!intl) {
    // Intl object not defined, fall back to returning default message
    return messageTemplate(message.defaultMessage, values);
  }

  return intl.formatMessage(message, values);
};

/**
 * Passes value to formatDate function of the intl object
 * @param {number} value - The timestamp to format
 */
export const formatDate = (value: number): string => intl.formatDate(value);

/**
 * Passes value to formatTime function of the intl object
 * @param {number} timestamp - The timestamp to format
 * @param {object} options - options to pass
 */
export const formatTime = (timestamp: number, options?: Object) =>
  intl.formatTime(timestamp, options);

export const getDisplayString = (message?: string | IntlMessage, values?: Object): string => {
  if (!message) return '';

  return typeof message === 'string' ? message : formatMessage(message, values);
};

// From https://stackoverflow.com/questions/377961/efficient-javascript-string-replacement
const messageTemplate = (s: string, values: {[string]: any}): string => {
  return s.replace(
    /{(\w*)}/g,
      (m: string, key: string) => {
        return values.hasOwnProperty( key ) ? values[key].toString() : "";
      }
    );
};