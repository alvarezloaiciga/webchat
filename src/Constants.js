export const MessageTypes = {
    TEXT: 'Text',
    CHAT_MESSAGE: 'ChatMessage',
};
