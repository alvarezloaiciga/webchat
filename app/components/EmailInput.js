// @flow
import React from 'react';
import styled from 'react-emotion';
import moment from 'moment-timezone';
import {css} from 'emotion';
import quiqOptions, {getStyle, getMessage} from 'Common/QuiqOptions';
import {isValidEmail} from 'Common/Utils';
import QuiqChatClient from 'quiq-chat';
import {darken} from 'polished';
import {UserEmailKey, messageTypes} from 'Common/Constants';

const EmailInputContainer = styled.div`
  display: flex;
  width: 100%;
  display: flex;
  align-items: center;
  overflow: auto;
  background: #fff;
`;

const ErrorStyle = css`box-shadow: inset 0px 0px 7px red;`;

const Input = styled.input`
  height: 100%;
  flex: 1 1 auto;
  border: 2px solid transparent;
  padding: 10px;
  border: none;
  font-size: 14px;
  line-height: 1.3em;
  margin-right: 10px;
  outline: none;
  user-select: none;
  ${props => (props.error ? ErrorStyle : '')};
`;

const IconCSS = (color: string) => css`
  flex: 0 0 auto;
  transition: 0.15s ease-in-out all;
  margin-right: 10px;
  color: ${color};
  cursor: pointer;

  &:hover {
    color: ${darken(0.15, color)};
  }
`;

const CancelButton = styled.i`${IconCSS('red')};`;

const SubmitButton = styled.i`${IconCSS('green')};`;

const {fontFamily, styles, host, contactPoint} = quiqOptions;

export type EmailInputProps = {
  onCancel: () => void,
  onSubmit: () => void, // eslint-disable-line react/no-unused-prop-types
};

type EmailInputState = {
  error: boolean,
  value: string,
};

const getInitialValue = () => {
  try {
    return atob(localStorage.getItem(`${UserEmailKey}_${contactPoint}`) || '');
  } catch (e) {
    return '';
  }
};

export class EmailInput extends React.Component<EmailInputProps, EmailInputState> {
  props: EmailInputProps;
  state: EmailInputState = {
    error: false,
    value: getInitialValue(),
  };
  input: any;

  componentDidMount() {
    // AutoFocus Input
    setTimeout(() => {
      if (this.input) this.input.focus();
      // Reset Input so cursor is at end of pre-populated input
      if (this.state.value) {
        const value = this.state.value;
        this.setState({value: ''}, () => {
          this.setState({value});
        });
      }
    }, 200);
  }

  submit = () => {
    if (isValidEmail(this.state.value)) {
      QuiqChatClient.emailTranscript({
        email: this.state.value.trim(),
        originUrl: host,
        timezone: moment.tz.guess(),
      });
      this.props.onSubmit();
    } else {
      this.setState({error: true});
    }
  };

  handleKeyDown = (e: SyntheticInputEvent<>) => {
    if (e.keyCode === 13) {
      this.submit();
    }
  };

  handleInputChange = (e: SyntheticInputEvent<>) => {
    this.setState({value: e.target.value});
  };

  render() {
    return (
      <EmailInputContainer
        style={getStyle(styles.EmailTranscriptInputContainer)}
        className="EmailInput"
      >
        <Input
          innerRef={r => {
            this.input = r;
          }}
          value={this.state.value}
          error={this.state.error}
          onChange={this.handleInputChange}
          onKeyDown={this.handleKeyDown}
          placeholder={getMessage(messageTypes.emailTranscriptInputPlaceholder)}
          style={getStyle(styles.EmailTranscriptInput, {
            fontFamily,
          })}
        />
        <CancelButton
          style={getStyle(styles.EmailTranscriptInputCancelButton)}
          title={getMessage(messageTypes.emailTranscriptInputCancelTooltip)}
          onClick={this.props.onCancel}
          className={`fa fa-times`}
        />
        <SubmitButton
          style={getStyle(styles.EmailTranscriptInputSubmitButton)}
          title={getMessage(messageTypes.emailTranscriptInputSubmitTooltip)}
          onClick={this.submit}
          data-test="submitButton"
          className={`fa fa-check`}
        />
      </EmailInputContainer>
    );
  }
}

export default EmailInput;
