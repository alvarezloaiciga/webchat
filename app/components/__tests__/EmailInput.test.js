// @flow
jest.mock('quiq-chat');
jest.mock('Common/QuiqOptions');

// @flow
import React from 'react';
import {UserEmailKey} from 'Common/Constants';
import type {EmailInputProps} from '../EmailInput';
import EmailInput from '../EmailInput';
import {shallow} from 'enzyme';
import type {ShallowWrapper} from 'enzyme';
import QuiqChatClient from 'quiq-chat';

describe('EmailInput component', () => {
  let wrapper: ShallowWrapper;
  let testProps: EmailInputProps;
  let render: () => void;
  QuiqChatClient.emailTranscript = jest.fn();

  beforeEach(() => {
    testProps = {
      onCancel: jest.fn(),
      onSubmit: jest.fn(),
    };
    render = () => {
      wrapper = shallow(<EmailInput {...testProps} />);
    };
  });

  describe('rendering', () => {
    beforeEach(() => {
      render();
    });

    it('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('input value', () => {
    it('retrieves the intial value from localStorage when available', () => {
      const email = 'andrew.jenkins@goquiq.com';
      localStorage.setItem(`${UserEmailKey}_Bob`, btoa(email));
      render();
      expect(wrapper.state('value')).toBe(email);
      localStorage.clear();
    });

    it('validates value submitted', () => {
      render();
      wrapper.setState({value: 'andrew.jenkins@goquiq.com'});
      wrapper.find('[data-test="submitButton"]').simulate('click');
      expect(testProps.onSubmit).toBeCalled();
    });

    it("doesn't call onsubmit with invalid emails", () => {
      render();
      wrapper.setState({value: 'invalidEmail'});
      wrapper.find('[data-test="submitButton"]').simulate('click');
      expect(testProps.onSubmit).not.toBeCalled();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
