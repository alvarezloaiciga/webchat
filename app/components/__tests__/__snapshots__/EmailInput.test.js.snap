// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailInput component input value doesn't call onsubmit with invalid emails 1`] = `
<Styled(div)
  className="EmailInput"
  style={
    Object {
      "border": "1px solid black",
    }
  }
>
  <Styled(input)
    error={true}
    innerRef={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Enter your Email..."
    style={
      Object {
        "fontFamily": "Lato, sans-serif",
        "outline": "none",
      }
    }
    value="invalidEmail"
  />
  <Styled(i)
    className="fa fa-times"
    onClick={[Function]}
    style={
      Object {
        "borderRadius": "10px",
      }
    }
    title="Cancel Email Transcript"
  />
  <Styled(i)
    className="fa fa-check"
    data-test="submitButton"
    onClick={[Function]}
    style={
      Object {
        "borderRadius": "10px",
      }
    }
    title="Email Transcript"
  />
</Styled(div)>
`;

exports[`EmailInput component rendering renders 1`] = `
<Styled(div)
  className="EmailInput"
  style={
    Object {
      "border": "1px solid black",
    }
  }
>
  <Styled(input)
    error={false}
    innerRef={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Enter your Email..."
    style={
      Object {
        "fontFamily": "Lato, sans-serif",
        "outline": "none",
      }
    }
    value=""
  />
  <Styled(i)
    className="fa fa-times"
    onClick={[Function]}
    style={
      Object {
        "borderRadius": "10px",
      }
    }
    title="Cancel Email Transcript"
  />
  <Styled(i)
    className="fa fa-check"
    data-test="submitButton"
    onClick={[Function]}
    style={
      Object {
        "borderRadius": "10px",
      }
    }
    title="Email Transcript"
  />
</Styled(div)>
`;
